File: D.S.A/arrays/medium/sliding-window/binary-subarray-with-sum.js
/**
 * @param {number[]} nums
 * @param {number} goal
 * @return {number}
 */
var numSubarraysWithSum = function(nums, goal) {
    

    let count =atmost(nums,goal)-atmost(nums,goal-1);

    return count;
};

 function atmost(nums,k){

        let r=0,l=0,count=0,sum=0;

        while(r<nums.length){
          
          if(k<0) return 0;

         sum+=nums[r];

         while(sum>k){
            sum-=nums[l];
            l++;
         }

         count+=r-l+1;

         r++;
        }

        return count;
    }

// Explanation:
// - Main idea: `exactly(goal) = atMost(goal) - atMost(goal - 1)`.
// - `atmost(nums, k)` counts subarrays whose sum is less than or equal to `k`.
// - Use sliding window with `l` and `r`, and keep current window `sum`.
// - Expand with `r`; if `sum > k`, shrink from left by moving `l`.
// - For each `r`, valid subarrays ending at `r` are `(r - l + 1)`, so add to `count`.
// - Subtracting the two atMost counts gives subarrays with sum exactly equal to `goal`.

File: Java/dsa/arrays/medium/sliding_window/BinarySubarrayWithSum.java
public class BinarySubarrayWithSum {
    public static int numSubarraysWithSum(int[] nums, int goal) {
        return atMost(nums, goal) - atMost(nums, goal - 1);
    }

    private static int atMost(int[] nums, int k) {
        if (k < 0) {
            return 0;
        }
        int l = 0;
        int sum = 0;
        int count = 0;

        for (int r = 0; r < nums.length; r++) {
            sum += nums[r];
            while (sum > k) {
                sum -= nums[l++];
            }
            count += r - l + 1;
        }

        return count;
    }
}
